### ISSUE TITLE
input_fn recall hook for estimator api

### ISSUE THREAD

**Issue by SumNeuron** [tensorflow/tensorflow#30046]
**System information**
- TensorFlow version (you are using): 1.10 < 2.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

Currently estimators are trained using either:

```python
estimator.train(...)
estimator.train_and_evaluate(...)    # internally calls self.train
```

Following the `tf.estimator.Estiamtor` [source code](https://github.com/tensorflow/estimator/blob/33e13754b0deebe7cc5006c49c53d246b6ff1dec/tensorflow_estimator/python/estimator/estimator.py) we see that `.train` calls `_train_model` that calls `_train_model_default`

[Line 1183](https://github.com/tensorflow/estimator/blob/33e13754b0deebe7cc5006c49c53d246b6ff1dec/tensorflow_estimator/python/estimator/estimator.py#L1183) has:

```python
# ...
features, labels, input_hooks = (self._get_features_and_labels_from_input_fn(input_fn, ModeKeys.TRAIN))
# ...
```

prior to calling `_train_with_estimator_spec`

with the features and labels from the `tf.data.Dataset` returned by the `input_fn` wrapped into an `EstimatorSpec`. 

So it seems that per epoch / per batch updates to the `tf.data.Dataset` is not possible without very careful construction of the dataset.

It would be nice to be able to have the `input_fn` have the ability to be recalled and possibly updated depending on estimator stated (e.g. depending on `step`)

**Will this change the current api? How?**

It _shouldn't_ change the current estimator api.
It may add some extra interfaces to the api (e.g. a decorator for the `input_fn` that passes `_step`, `_recall_on_batch`)

**Who will benefit with this feature?**

Users with data that requires run-time generation / variation 

**Any Other info.**

e.g. this kind of logic should be possible
```python
def input_fn(..., _step, _epoch):

    if _step > n:
        return tf.data.Dataset(...)
    return tf.data.Dataset(...)
```

**Comment by rmothukuru** [comment]
@SumNeuron,
Sorry for the delayed response. In the **`Tensorflow Version 2.x`**, since we use [TF Keras](https://www.tensorflow.org/api_docs/python/tf/keras/) and [tf.data](https://www.tensorflow.org/guide/data) predominantly and don't use [Estimators](https://www.tensorflow.org/guide/estimator) much, can you please let us know if this Feature is still relevant? 

Thanks!
